### DISCUSSION TITLE
How to test and validate a template locally?

### DISCUSSION THREAD

**Discussion by jsjoeio** [coder/coder#5452] [General] [answered]
I made a couple changes to the dogfooding template. How do I test and validate locally using `terraform`?

**Comment by jsjoeio** [comment, marked as answer]
@johnstcn showed me his workflow:

1. make a temporary directory: `mkdir tmp/tf-template-test`
2. inside that dir: `touch main.tf` and add your Terraform code
3. run `terraform init`
4. run `terraform apply`

Here's an example for some code that creates a script:

`main.tf`

```terraform
variable "dotfiles_uri" {
  type        = string
}

resource "local_file" "script" {
  filename = "${path.module}/script.sh"
  content  = <<-EOF
    #!/bin/sh
    set -x
    # install and start code-server
    curl -fsSL https://code-server.dev/install.sh | sh -s -- --version 4.8.3
    code-server --auth none --port 13337 &
    sudo service docker start
    DOTFILES_URI=${var.dotfiles_uri}
    if [ -n "$DOTFILES_URI" ]; then
      coder dotfiles "$DOTFILES_URI" -y 2>&1 | tee  ~/.personalize.log
    fi
	EOF
}

provider "local" {}
```

**How do I pass a value if the variable is optional?**

Like this: 
```shell
terraform apply -var dotfiles_uri="https://github.com/jsjoeio/dotfiles.git"
```

**Comment by dolgeneraldgme** [comment]
Valid template
The following example passes the --template-url parameter, to validate a template in a remote location:


aws cloudformation validate-template --template-url s3.amazonaws.com/cloudformation-templates-us-east-1/S3_Bucket.template
The following is the output of the previous command.

{
    "Description": "AWS CloudFormation Sample Template S3_Bucket: Sample template showing how to create a publicly accessible S3 bucket. **WARNING** This template creates an S3 bucket.
You will be billed for the AWS resources used if you create a stack from this template.",
    "Parameters": [],
    "Capabilities": []
}
The expected result is no error message, with information about all parameters listed.

Validation error
This section produces a validation error when running the aws cloudformation validate-template command.

The following example passes the --template-body parameter, to validate a template locally.


aws cloudformation validate-template --template-body file:///home/local/test/sampletemplate.json
The following is the output of the previous command.

{
    "ResponseMetadata": {
        "RequestId": "4ae33ec0-1988-11e3-818b-e15a6df955cd"
    },
    "Errors": [
        {
            "Message": "Template format error: JSON not well-formed. (line 11, column 8)",
            "Code": "ValidationError",
            "Type": "Sender"
        }
    ],
    "Capabilities": [],
    "Parameters": []
}
A client error (ValidationError) occurred: Template format error: JSON not well-formed. (line 11, column 8)
The expected result is an error message, with information about error listed.
Please reach out here for further details [https://dolgeneraldgme.com](url)
